File: scala/Eithers.scala
import Spec._
import Undefined._

object Eithers {

  def failIf[A](p: Boolean, a: => A): Either[A, Unit] =
    if (p) Left(a) else Right(())

  def getUser(req: Request): Either[Response, User] = for {
    token <- req.header.get("Authorization").toRight(noToken).right
    _     <- failIf(`malformed token?`, malformedToken(token)).right
    _     <- failIf(`user not found?`, noUser(token)).right
  } yield `the user`

  def getResource(req: Request): Either[Response, Resource] = for {
    path <- Right(req.path).right
    _    <- failIf(`resource not found?`, noResource(path)).right
  } yield `the resource`

  def execute(content: String, usr: User, src: Resource):
  Either[Response, Unit] = for {
    _ <- failIf(! `is permitted?`, notPermitted(src.path)).right
    _ <- failIf(! `is executed?`, badConnection).right
  } yield {}

  def handlePost(req: Request): Response = {
    for {
      _   <- failIf(req.method != "POST", notAllowed(req.method)).right
      _   <- failIf(req.body.isEmpty, noBody).right
      usr <- getUser(req).right
      src <- getResource(req).right
      _   <- execute(req.body, usr, src).right
    } yield success(req.path, req.body)
  }.fold(identity, identity)
}

File: haskell/Eithers.hs
module Eithers where

import Spec
import Control.Monad (join)

implies :: Bool -> e -> Either e ()
failure `implies` fallback = if failure then Left fallback else Right ()

implies' :: IO Bool -> e -> IO (Either e ())
failures `implies'` fallback = (`implies` fallback) <$> failures

tunnel :: Either e (IO (Either e a)) -> IO (Either e a)
tunnel eitherIoEither = join <$> sequenceA eitherIoEither

getUser :: Request -> Either Response User
getUser (Request _ _ _ header) = do
  token <- maybe (Left noToken) Right $ lookup "Authorization" header
  is_malformed_token `implies` malformedToken token
  is_user_not_found `implies` noUser token
  return the_user

getResource :: () -> Request -> IO (Either Response Resource)
getResource method (Request path _ _ _) = do
  let doResource = (\_ -> the_resource_io) :: () -> IO Resource
  notFound <- is_resource_not_found_io `implies'` noResource path
  doResource `traverse` notFound

execute :: () -> String -> User -> Resource -> IO (Either Response ())
execute method body usr (Resource path) = do
  let doExecuted = (\_ -> is_executed_io) :: () -> IO Bool
  permitted <- (not <$> is_permitted_io) `implies'` notPermitted path
  executed <- doExecuted `traverse` permitted
  return ((not <$> executed) >>= (`implies` badConnection))

handlePost :: Request -> IO Response
handlePost req@(Request path method body _) = either id id <$> result where
  chkMth = (method /= "POST") `implies` notAllowed method
  errBdy = null body `implies` noBody >> return body
  errUsr = getUser req
  precon = chkMth >> errBdy >> errUsr >> return ()
  result = do
    errSrc <- tunnel (getResource <$> precon <*> pure req)
    errExe <- tunnel (execute <$> precon <*> errBdy <*> errUsr <*> errSrc)
    return (errExe >> return (success path body))
